File: 131-palindrome-partitioning/palindrome-partitioning.ts
function partition(s: string): string[][] {
    const res: string[][] = [];
    const path: string[] = [];

    function isPalindrome(str: string, start: number, end: number): boolean {
        while (start <= end) {
            if (str[start] !== str[end]) {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }

    function partitionHelper(index: number): void {
        if (index === s.length) {
            res.push([...path]);
            return;
        }

        for (let i = index; i < s.length; i++) {
            if (isPalindrome(s, index, i)) {
                path.push(s.substring(index, i + 1));
                partitionHelper(i + 1);
                path.pop();
            }
        }
    }

    partitionHelper(0);
    return res;
}

File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    List<List<String>> ans = new ArrayList<>();
    List<String> curList = new ArrayList<>();

    public List<List<String>> partition(String s) {
        helper(0, s);
        return ans;
    }

   private void helper(int curIndex, String s){
    if(curIndex == s.length()){
        ans.add(new ArrayList(curList));
        return;
    }

    for(int i = curIndex ; i < s.length(); i ++){
        if(isPalindrome(s , curIndex , i)){
            curList.add(s.substring(curIndex , i + 1));
            helper(i + 1 , s);
            curList.remove(curList.size() - 1);
        } 
    }

   }

    private boolean isPalindrome(String s, int start, int end) {
        while (start <= end) {
            if (s.charAt(start) != s.charAt(end)) {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
}